pkg/authorization/delegated/caching_authorizer.go

```diff
@@ -0,0 +1,134 @@
+	tenancyv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/tenancy/v1alpha1"
+	dynamiccontext "github.com/kcp-dev/kcp/pkg/virtual/framework/dynamic/context"
+)
+
+type CachingOptions struct {
+	Options
+
+	// Verb is the verb being authorized, passed in authorizer.AttributesRecord.
+	Verb string
+
+	// Resource is the resource being authorized, passed in authorizer.AttributesRecord.
+	Resource string
+
+	// TTL is the default time-to-live when a delegated authorizer
+	// is stored in the internal cache.
+	TTL time.Duration
+}
+
+func (c *CachingOptions) defaults() {
+	c.Options.defaults()
+	if c.TTL == 0 {
+		c.TTL = 12 * time.Hour
+	}
+}
+
+// NewCaching creates a new Authorizer that holds an internal cache of
+// Delegated Authorizer(s).
+// A Delegated Authorizer, in turn, caches Authorize requests for the
+func NewCaching(client kcpkubernetesclientset.ClusterInterface, opts CachingOptions) *cachingAuthorizer {
+	opts.defaults()
+	return &cachingAuthorizer{
+		opts:   &opts,
+		cache:  cache.NewExpiring(),
+		client: client,
+	}
+}
+
+// cachingAuthorizer is a wrapper around authorizer.Authorize that uses
+// an internal expiring cache.
+type cachingAuthorizer struct {
+	opts  *CachingOptions
+	cache *cache.Expiring
+
+	client kcpkubernetesclientset.ClusterInterface
+}
+
+// load loads the authorizer from the cache, if any.
+func (c *cachingAuthorizer) load(clusterName logicalcluster.Name) authorizer.Authorizer {
+	value, ok := c.cache.Get(clusterName)
+	if !ok && value == nil {
+		return nil
+	}
+	authz, ok := value.(authorizer.Authorizer)
+	if !ok {
+		return nil
+	}
+	return authz
+}
+
+func (c *cachingAuthorizer) loadOrStore(clusterName logicalcluster.Name) (authorizer.Authorizer, error) {
+	if authz := c.load(clusterName); authz != nil {
+		return authz, nil
+	}
+
+	// Create the delegated authorizer.
+	authz, err := NewDelegatedAuthorizer(clusterName, c.client, c.opts.Options)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store the cache and return.
+	c.cache.Set(clusterName, authz, c.opts.TTL)
+	return authz, nil
+}
+
+func (c *cachingAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	logger := klog.FromContext(ctx)
+	clusterName, name, err := initialization.TypeFrom(
+		corev1alpha1.LogicalClusterInitializer(dynamiccontext.APIDomainKeyFrom(ctx)),
+	)
+	if err != nil {
+		logger.V(2).Info(err.Error())
+		return authorizer.DecisionNoOpinion, "unable to determine initializer", fmt.Errorf("access not permitted")
+	}
+
+	authz, err := c.loadOrStore(clusterName)
+	if err != nil {
+		return authorizer.DecisionNoOpinion, "error", err
+	}
+
+	SARAttributes := authorizer.AttributesRecord{
+		APIGroup:        tenancyv1alpha1.SchemeGroupVersion.Group,
+		APIVersion:      tenancyv1alpha1.SchemeGroupVersion.Version,
+		User:            attr.GetUser(),
+		Name:            name,
+		ResourceRequest: c.opts.Resource != "",
```

Comment:
Is this reasonable to do? @s-urbaniak

Reply:
In alternative, we could just expose this in the Options?

Reply:
@vincepri thinking 🤔 how about, for reuse, we leave the cache simply being a cache and not let it implement the `Authorize` k8s interface. That way consumers of the cached authorizer can refactor to use the cache in a minimal way. Instead of instantiating using `NewDelegatedAuthorizer`, they simply invoke a `GetDelegatedAuthorizer(clusterName)` method on the cache and leave existing logic as-is. Here is a commit where I did the refactoring to reduce the amount of comments here: https://github.com/kcp-dev/kcp/commit/99fbf479a9f96ab023306c7602bef1b5f698cb52

Using this the diff in the initiliazingworkspaces builder is minimal and boils down to:
```

+       cachingAuthorizer := newAuthorizer(kubeClusterClient, delegated.NewDelegatedAuthorizerCache(kubeClusterClient, delegated.CachingOptions{}))

...

-func newAuthorizer(client kcpkubernetesclientset.ClusterInterface) authorizer.AuthorizerFunc {
+func newAuthorizer(client kcpkubernetesclientset.ClusterInterface, authorizerCache *delegated.DelegatedAuthorizerCache) authorizer.AuthorizerFunc {
        return func(ctx context.Context, attr authorizer.Attributes) (authorizer.Decision, string, error) {
                clusterName, name, err := initialization.TypeFrom(corev1alpha1.LogicalClusterInitializer(dynamiccontext.APIDomainKeyFrom(ctx)))
                if err != nil {
                        return authorizer.DecisionNoOpinion, "unable to determine initializer", fmt.Errorf("access not permitted")
                }

-               authz, err := delegated.NewDelegatedAuthorizer(clusterName, client)
+               authz, err := authorizerCache.GetDelegatedAuthorizer(clusterName)
                if err != nil {
                        return authorizer.DecisionNoOpinion, "error", err
                }
```

Raison d'etre:
- Authorizers derive the cluster name using different ways, i.e. some use a specific context (like here), some derive it from the request path
- The subject access review attributes are very different between authorizers

WDYT?